src/uct/rocm/rocm_common.c

```diff
@@ -0,0 +1,261 @@
+
+
+/** Callback to enumerate pools for given agent.
+ *  Try to find global pool assuming one global pool per agent.
+*/
+static hsa_status_t uct_rocm_hsa_amd_memory_pool_callback(
+                                hsa_amd_memory_pool_t memory_pool, void* data)
+{
+    hsa_status_t status;
+    hsa_amd_segment_t amd_segment;
+
+    status = hsa_amd_memory_pool_get_info(memory_pool,
+                                         HSA_AMD_MEMORY_POOL_INFO_SEGMENT,
+                                         &amd_segment);
+
+    if (status != HSA_STATUS_SUCCESS) {
+        ucs_error("Failure to get pool info: 0x%x", status);
+        return status;
+    }
+
+    if (amd_segment ==  HSA_AMD_SEGMENT_GLOBAL) {
+        *(hsa_amd_memory_pool_t *)data = memory_pool;
+        ucs_debug("Found global pool: 0x%lx", memory_pool.handle);
+        return HSA_STATUS_INFO_BREAK;
+    }
+
+    return HSA_STATUS_SUCCESS;
+}
+
+/** Callback to enumerate HSA agents */
+static hsa_status_t uct_rocm_hsa_agent_callback(hsa_agent_t agent, void* data)
+{
+    uint32_t bdfid;
+    hsa_device_type_t device_type;
+    hsa_status_t status;
+
+    ucs_debug("hsa_agent_callback: Agent  0x%lx", agent.handle);
+
+    status = hsa_agent_get_info(agent, HSA_AGENT_INFO_DEVICE, &device_type);
+
+    if (status != HSA_STATUS_SUCCESS) {
+        ucs_error("Failure to get device type: 0x%x", status);
+        return status;
+    }
+
+    if (device_type == HSA_DEVICE_TYPE_GPU) {
+
+        status = hsa_agent_get_info(agent, HSA_AMD_AGENT_INFO_BDFID, &bdfid);
+
+        if (status != HSA_STATUS_SUCCESS) {
+            ucs_warn("Failure to get pci info: 0x%x", status);
+            return status;
+        }
+
+        uct_rocm_cfg.agents.gpu_agent[uct_rocm_cfg.num_of_gpu] = agent;
+        uct_rocm_cfg.agents.gpu_info[uct_rocm_cfg.num_of_gpu].bus = (bdfid >> 8) & 0xff;
+        uct_rocm_cfg.agents.gpu_info[uct_rocm_cfg.num_of_gpu].device = (bdfid >> 3) & 0x1F;
+        uct_rocm_cfg.agents.gpu_info[uct_rocm_cfg.num_of_gpu].func = bdfid & 0x7;
+
+        ucs_debug("Found GPU agent : 0x%lx. [ B#%02d, D#%02d, F#%02d ]",
+                uct_rocm_cfg.agents.gpu_agent[uct_rocm_cfg.num_of_gpu].handle,
+                uct_rocm_cfg.agents.gpu_info[uct_rocm_cfg.num_of_gpu].bus,
+                uct_rocm_cfg.agents.gpu_info[uct_rocm_cfg.num_of_gpu].device,
+                uct_rocm_cfg.agents.gpu_info[uct_rocm_cfg.num_of_gpu].func);
+
+
+        uct_rocm_cfg.agents.gpu_info[uct_rocm_cfg.num_of_gpu].pool.handle
+                                                            = (uint64_t) -1;
+        status = hsa_amd_agent_iterate_memory_pools(agent,
+                                    uct_rocm_hsa_amd_memory_pool_callback,
+                                    &uct_rocm_cfg.agents.gpu_info[uct_rocm_cfg.num_of_gpu].pool);
+
+        if (status != HSA_STATUS_SUCCESS && status != HSA_STATUS_INFO_BREAK) {
+            ucs_error("Failure to iterate regions: 0x%x\n", status);
+            return status;
```

Comment:
About cleanup: We only failed to collect info (query failed) so nothing to clean-up: no special resources were requested, etc. Sp only error status will be returned back to upper layer.